### PR TITLE
autolinks system tiddlers

### PR DESCRIPTION
**Pull request by tobibeer**
Implements #1103 — **Autolink system tiddler references**

* enhanced extlink parser to cover system tiddler links
    * of course, only those without spaces in their title
    * didn't want to open up an extra parser for this as extlink has all the basics

### CODE REVIEW

**Comment by Jermolene** [comment]
Hi @tobibeer many thanks. I think that reusing extlink.js to parse internal links to system tiddlers is a bit confusing. I'd prefer a separate parse rule.

**Comment by tobibeer** [comment]
done

File: core/modules/parsers/wikiparser/rules/syslink.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+/*global $tw: false */
+"use strict";
+
+exports.name = "syslink";
+exports.types = {inline: true};
+
+exports.init = function(parser) {
+	this.parser = parser;
+	// Regexp to match
+	this.matchRegExp = /~?(?:\$):[^\s<>{}\[\]`|'"\\^~]+(?:\/|\b)/mg;
```
Comment (pmario) [inline core/modules/parsers/wikiparser/rules/syslink.js]: hmmm: `$:/asdf{[]}` imo is a valid system tiddler name? ... at least creating it isn't prevented.

The excluded characters may be illegal with the external link regexp. But imo for system tiddlers they aren't.


full regexp description
```
// ~?(?:\$):[^\s<>{}\[\]`|'"\\^~]+(?:\/|\b)
// 
// Options: case insensitive; ^ and $ match at line breaks
// 
// Match the character “~” literally «~?»
//    Between zero and one times, as many times as possible, giving back as needed (greedy) «?»
// Match the regular expression below «(?:\$)»
//    Match the character “$” literally «\$»
// Match the character “:” literally «:»
// Match a single character NOT present in the list below «[^\s<>{}\[\]`|'"\\^~]+»
//    Between one and unlimited times, as many times as possible, giving back as needed (greedy) «+»
//    A whitespace character (spaces, tabs, and line breaks) «\s»
//    One of the characters “<>{}” «<>{}»
//    A [ character «\[»
//    A ] character «\]»
//    One of the characters “`|'"” «`|'"»
//    A \ character «\\»
//    One of the characters “^~” «^~»
// Match the regular expression below «(?:\/|\b)»
//    Match either the regular expression below (attempting the next alternative only if this one fails) «\/»
//       Match the character “/” literally «\/»
//    Or match regular expression number 2 below (the entire group fails if this one fails to match) «\b»
//       Assert position at a word boundary «\b»
```

Comment (tobibeer) [inline core/modules/parsers/wikiparser/rules/syslink.js]: Maybe this: `this.matchRegExp = /~?(?:\$):[^\s|]+(?:\/|\b)/mg;` ?

Comment (pmario) [inline core/modules/parsers/wikiparser/rules/syslink.js]: I'm not sure. @Jermolene, @Skeeve ??

Comment (tobibeer) [inline core/modules/parsers/wikiparser/rules/syslink.js]: The above is not good enough as it destroys system tiddler titles.

Comment (Jermolene) [inline core/modules/parsers/wikiparser/rules/syslink.js]: Are there any core system tiddlers that contain angle brackets?

**Comment by Jermolene** [comment]
I'd be thinking that the regexp should match all of the predefined core system tiddlers. There are no restrictions on system tiddler titles besides starting with `$:/`, so we shouldn't be trying to match all legal system tiddler titles.

**Comment by pmario** [comment]
Then we can go with Tobias's first version. ... I had a closer look and it seems to be fine for the existing system tiddlers. ... all kind of brackets in tiddler names, may cause problems as widget / macro parameters anyway.

**Comment by Jermolene** [comment]
I think that's fine; this is really a convenience for simple system tiddlers like `$:/ControlPanel`, `$:/core/ui/PageTemplate` etc.

File: core/modules/parsers/wikiparser/rules/syslink.js (hunk 1 above)
Comment (tobibeer) [inline core/modules/parsers/wikiparser/rules/syslink.js]: > Are there any core system tiddlers that contain angle brackets?

Nope, but the template that renders the tiddler title breaks for system tiddlers if angle brackets are allowed in this regexp.

Comment (Jermolene) [inline core/modules/parsers/wikiparser/rules/syslink.js]: > Nope, but the template that renders the tiddler title breaks for system tiddlers if angle brackets are allowed in this regexp.

Do you mean `$:/core/ui/ViewTemplate/title`? Where does it break?

Comment (tobibeer) [inline core/modules/parsers/wikiparser/rules/syslink.js]: Didn't further inspect where something breaks. I just noticed that when looking at a system tiddler, the title would be rendered with a bunch of broken html with the first simplified regexp above, so I readded those angle brackets to the regexp which fixed that problem.

Comment (Jermolene) [inline core/modules/parsers/wikiparser/rules/syslink.js]: Do you mean the rendering of the tiddler title from `$:/core/ui/ViewTemplate/title`? I don't understand how that could be affected by the regexp used here.

Comment (tobibeer) [inline core/modules/parsers/wikiparser/rules/syslink.js]: Me neither. If you find this interesting to explore, please do.
I wouldn't quite know where to start looking.

Comment (Jermolene) [inline core/modules/parsers/wikiparser/rules/syslink.js]: > Me neither. If you find this interesting to explore, please do.

It's a blocker for accepting this pull request. It means that the pull request has a failure mode that we don't fully understand. That makes it risky to accept.

Comment (tobibeer) [inline core/modules/parsers/wikiparser/rules/syslink.js]: > It's a blocker for accepting this pull request. It means that the pull request has a failure mode that we don't fully understand. That makes it risky to accept.

Just to make sure, the pull request has no known failure mode.
What does have a failure mode is a non-working "simplification" of it.

Comment (Jermolene) [inline core/modules/parsers/wikiparser/rules/syslink.js]: > What does have a failure mode is a non-working "simplification" of it.

But we have no explanation for how this regexp affects the display of the title; that means that we don't understand how this regexp works.

Comment (Skeeve) [inline core/modules/parsers/wikiparser/rules/syslink.js]: Am 10.01.15 15:40, schrieb Tobias Beer - notifications@github.com:
> this.matchRegExp = /~?(?:\$):[^\s<>{}\[\]`|'"\\^~]+(?:\/|\b)/mg
~? => Fine
(?:\$) => reducable to \$
: => fine
[^\s<>{}\[\]`|'"\\^~] => Okay, all of those characters are excluded and 
may not occure anywhere in the link
(?:\/|\b) => So the end is either at a / or a word boundary? I think it 
should be reducable to \b as / already denotes a word boundary if 
followed by a word character.

Anyhow: I think I do not fully understand what this is about.

Comment (Jermolene) [inline core/modules/parsers/wikiparser/rules/syslink.js]: > Anyhow: I think I do not fully understand what this is about

I think that @tobibeer is saying that when he experimentally removed angled brackets from the regexp for matching system tiddlers then he experienced a crash when viewing a system tiddler that had angle brackets in its title.

**Comment by Jermolene** [comment]
> Then we can go with Tobias's first version.

Does that correspond with the current state of this pull request?

File: core/modules/parsers/wikiparser/rules/syslink.js (hunk 1 above)
Comment (tobibeer) [inline core/modules/parsers/wikiparser/rules/syslink.js]: Not a crash (no js error), only the titles are jumbled up...

jumbled

So, this doesn't work...

`this.matchRegExp = /~?\$:[^\s|]+(?:\/|\b)/mg;`

This works...

`this.matchRegExp = /~?\$:[^\s<>|]+(?:\/|\b)/mg;`

**Comment by tobibeer** [comment]
>> Then we can go with Tobias's first version.

> Does that correspond with the current state of this pull request?

I think yes, the regexp may be a wee bit too restrictive, but I think that's fine for now...

As mentioned above, a bit less restrictive, this appears to work as well...

```
this.matchRegExp = /~?\$:[^\s<>|]+(?:\/|\b)/mg;
```

As compared to the current...

```
this.matchRegExp = /~?(?:\$):[^\s<>{}\[\]`|'"\\^~]+(?:\/|\b)/mg;
```

**Comment by pmario** [comment]
@tobibeer what happens if you use `[[$:/asdf/asdf]]`? ... Does the rendering work?  Because there are a lot system links like this in the existing documentation.

**Comment by tobibeer** [comment]
> What happens if you use [[$:/asdf/asdf]]? ... Does the rendering work?

It does. the precedence order for parsers must be handled elsewhere so as to ensure it does.

**Comment by tobibeer** [comment]
Ok, I think I've got it now. The following matches all...

1. starting with a literal `$:`
2. then any number of character not a whitespace, `<` or  `|`
3. closing with anything that is, again, not a whitespace, `<` or  `|`

Test here...

http://tbdemo.tiddlyspot.com/#Autolink%20System%20Tiddlers

**Comment by Skeeve** [comment]
I think, this test is irrelevant:

|[[$:/foo/bar]] — [[$:/foo/bar/]]|


|As your changes are never hit by that.


Try putting spaces or in:


|

|[[$:/foo/bar baz]]

It'll still link.
|

||

**Comment by tobibeer** [comment]
I know, I have only put it into the examples because Mario expressed a little concern over all the existing examples... so it would be clear that this is not affected. Although, I'd be interested in understanding how the precedence of parsers is established.

**Comment by Skeeve** [comment]
Am 11.01.15 11:59, schrieb Tobias Beer - notifications@github.com:
>
> Although, I'd be interested in understanding how the precedence of 
> parsers is established.
>
Did you take a look at $:/core/modules/parsers/wikiparser/wikiparser.js ?

It say in the comments:

|Get the next match out of an array of parse rule instances
*/
WikiParser.prototype.findNextMatch = function(rules,startPos) {
	// Find the best matching rule by finding the closest match position|

**Comment by tobibeer** [comment]
> Did you take a look at $:/core/modules/parsers/wikiparser/wikiparser.js ?

Thanks a lot for the pointer, much appreciated ...and, no.
I prefer exploring the core code on an as-needed basis...
mostly because it's hard to otherwise figure out
the general narrative just by looking at individual modules.

**Pull request merged**